File: src/stores/ssh.ts
import { create } from "zustand";
import {
  sshAssignProfile,
  sshConfigOpen,
  sshGenerate,
  sshImport,
  sshList,
  sshRemove,
  sshReveal,
  sshRevealFolder,
  sshScan,
} from "@/ipc";
import { type AppError, toAppError } from "@/ipc/errors";
import type { SshAlgorithm, SshKey } from "@/ipc/types.gen";
import { useActivityStore } from "@/stores/activity";
import { useFeedbackStore } from "@/stores/feedback";

export interface GenerateInput {
  label: string;
  algorithm: SshAlgorithm;
  comment: string | null;
  outputDir: string;
  fileName: string;
  hostAlias: string | null;
  hostName: string | null;
}

interface SshStore {
  items: SshKey[];
  loading: boolean;
  error: AppError | null;
  importOpen: boolean;
  generateOpen: boolean;
  setImportOpen: (open: boolean) => void;
  setGenerateOpen: (open: boolean) => void;
  fetch: () => Promise<void>;
  scan: () => Promise<void>;
  importKey: (
    path: string,
    hostAlias?: string | null,
    hostName?: string | null
  ) => Promise<boolean>;
  generate: (input: GenerateInput) => Promise<boolean>;
  remove: (id: string) => Promise<boolean>;
  reveal: (id: string) => Promise<void>;
  revealFolder: () => Promise<void>;
  openConfig: () => Promise<void>;
  assignProfile: (id: string, profileId: string | null) => Promise<void>;
}

export const useSshStore = create<SshStore>((set, get) => ({
  items: [],
  loading: false,
  error: null,
  importOpen: false,
  generateOpen: false,

  setImportOpen: (open) => set({ importOpen: open }),
  setGenerateOpen: (open) => set({ generateOpen: open }),

  fetch: async () => {
    set({ loading: true, error: null });
    try {
      const items = await sshList();
      set({ items, loading: false });
    } catch (e) {
      set({ error: toAppError(e), loading: false });
    }
  },

  scan: async () => {
    set({ loading: true, error: null });
    try {
      const items = await sshScan();
      set({ items, loading: false });
    } catch (e) {
      const err = toAppError(e);
      set({ error: err, loading: false });
      useFeedbackStore.getState().toast(err.message, "error");
    }
  },

  importKey: async (path, hostAlias, hostName) => {
    try {
      await sshImport(path, hostAlias ?? null, hostName ?? null);
      await get().fetch();
      useFeedbackStore.getState().toast("SSH key imported", "success");
      useActivityStore.getState().fetch();
      return true;
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
      return false;
    }
  },

  generate: async (input) => {
    try {
      await sshGenerate(
        input.label,
        input.algorithm,
        input.comment,
        input.outputDir,
        input.fileName,
        input.hostAlias,
        input.hostName
      );
      await get().fetch();
      useFeedbackStore.getState().toast("SSH key generated", "success");
      useActivityStore.getState().fetch();
      return true;
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
      return false;
    }
  },

  remove: async (id) => {
    try {
      await sshRemove(id);
      await get().fetch();
      useFeedbackStore.getState().toast("SSH key removed", "success");
      useActivityStore.getState().fetch();
      return true;
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
      return false;
    }
  },

  reveal: async (id) => {
    try {
      await sshReveal(id);
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
    }
  },

  revealFolder: async () => {
    try {
      await sshRevealFolder();
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
    }
  },

  openConfig: async () => {
    try {
      await sshConfigOpen();
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
    }
  },

  assignProfile: async (id, profileId) => {
    try {
      await sshAssignProfile(id, profileId);
      await get().fetch();
      useFeedbackStore
        .getState()
        .toast(profileId ? "Key assigned to profile" : "Key unassigned", "success");
      useActivityStore.getState().fetch();
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
    }
  },
}));

File: src-tauri/src/commands/ssh.rs
use crate::{
    domain::ssh::{GenerateParams, SshAlgorithm, SshKey},
    error::AppError,
    infra::paths,
    state::AppState,
};
use tauri::State;
use tauri_plugin_opener::OpenerExt;
use uuid::Uuid;

#[tauri::command]
pub(crate) async fn ssh_list(state: State<'_, AppState>) -> Result<Vec<SshKey>, AppError> {
    state.ssh.list()
}

#[tauri::command]
pub(crate) async fn ssh_scan(state: State<'_, AppState>) -> Result<Vec<SshKey>, AppError> {
    let dir = paths::ssh_dir()?;
    state.ssh.scan(&dir)
}

#[tauri::command]
pub(crate) async fn ssh_import(
    private_key_path: String,
    host_alias: Option<String>,
    host_name: Option<String>,
    state: State<'_, AppState>,
) -> Result<SshKey, AppError> {
    if private_key_path.trim().is_empty() {
        return Err(AppError::Validation("no key file selected".into()));
    }
    state.ssh.import(private_key_path, host_alias, host_name)
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub(crate) async fn ssh_generate(
    label: String,
    algorithm: SshAlgorithm,
    comment: Option<String>,
    output_dir: String,
    file_name: String,
    host_alias: Option<String>,
    host_name: Option<String>,
    state: State<'_, AppState>,
) -> Result<SshKey, AppError> {
    state.ssh.generate(GenerateParams {
        label,
        algorithm,
        comment,
        output_dir,
        file_name,
        host_alias,
        host_name,
    })
}

#[tauri::command]
pub(crate) async fn ssh_remove(id: Uuid, state: State<'_, AppState>) -> Result<(), AppError> {
    state.ssh.remove(id)
}

#[tauri::command]
pub(crate) async fn ssh_assign_profile(
    id: Uuid,
    profile_id: Option<Uuid>,
    state: State<'_, AppState>,
) -> Result<SshKey, AppError> {
    state.ssh.assign_profile(id, profile_id)
}

#[tauri::command]
pub(crate) async fn ssh_reveal(
    id: Uuid,
    app: tauri::AppHandle,
    state: State<'_, AppState>,
) -> Result<(), AppError> {
    // Only reveal a validated, stored key path — never arbitrary input.
    let key = state.ssh.mark_used(id)?;
    app.opener()
        .reveal_item_in_dir(&key.private_key_path)
        .map_err(|e| AppError::Io(e.to_string()))
}

#[tauri::command]
pub(crate) async fn ssh_reveal_folder(app: tauri::AppHandle) -> Result<(), AppError> {
    let dir = paths::ssh_dir()?;
    app.opener()
        .reveal_item_in_dir(&dir)
        .map_err(|e| AppError::Io(e.to_string()))
}

#[tauri::command]
pub(crate) async fn ssh_config_open(app: tauri::AppHandle) -> Result<(), AppError> {
    // Opens only the fixed ~/.ssh/config path in the default editor.
    let path = paths::ssh_config_path()?;
    app.opener()
        .open_path(path.to_string_lossy(), None::<&str>)
        .map_err(|e| AppError::Io(e.to_string()))
}

#[tauri::command]
pub(crate) async fn ssh_config_preview(state: State<'_, AppState>) -> Result<String, AppError> {
    state.ssh.config_preview()
}
